featurebyte/service/target.py

```diff
@@ -14,3 +70,129 @@ class TargetService(BaseDocumentService[TargetModel, TargetCreate, TargetService
+        )
+        node = graph.get_node_by_name(node_name_map[target.node_name])
+
+        # prune the graph to remove unused nodes
+        pruned_graph, pruned_node_name_map = QueryGraph(**graph.dict(by_alias=True)).prune(
+            target_node=node
+        )
+        if sanitize_for_definition:
+            pruned_graph = sanitize_query_graph_for_feature_definition(graph=pruned_graph)
+        return pruned_graph, pruned_node_name_map[node.name]
+
+    async def _get_target_version(self, name: str) -> VersionIdentifier:
+        version_name = get_version()
+        query_result = await self.list_documents(
+            query_filter={"name": name, "version.name": version_name}
+        )
+        count = query_result["total"]
+        return VersionIdentifier(name=version_name, suffix=count or None)
+
+    async def create_document(self, data: TargetCreate) -> TargetModel:
+        document = await self.prepare_target_model(data=data, sanitize_for_definition=False)
+        async with self.persistent.start_transaction() as session:
+            # check any conflict with existing documents
+            await self._check_document_unique_constraints(document=document)
+
+            # insert the document
+            params = document.dict(by_alias=True)
+            if data.graph:
+                params["raw_graph"] = data.graph.dict()
+            insert_id = await session.insert_one(
+                collection_name=self.collection_name,
+                document=params,
+                user_id=self.user.id,
+            )
+            assert insert_id == document.id
+
+            try:
+                target_namespace = await self.target_namespace_service.get_document(
+                    document_id=document.target_namespace_id,
+                )
+                await validate_target_version_and_namespace_consistency(
+                    target=document, target_namespace=target_namespace
+                )
+                await self.target_namespace_service.update_document(
+                    document_id=document.target_namespace_id,
+                    data=TargetNamespaceUpdate(
+                        target_ids=self.include_object_id(target_namespace.target_ids, document.id)
+                    ),
+                    return_document=True,
+                )
+            except DocumentNotFoundError:
+                entity_ids = document.entity_ids or []
+                await self.target_namespace_service.create_document(
+                    data=TargetNamespaceCreate(
+                        _id=document.target_namespace_id,
+                        name=document.name,
+                        target_ids=[insert_id],
+                        default_target_id=insert_id,
+                        default_version_mode=DefaultVersionMode.AUTO,
+                        entity_ids=sorted(entity_ids),
+                    ),
+                )
+        return await self.get_document(document_id=insert_id)
```

Comment:
Seems like this part shares many common logics with `FeatureService` class. Would be good if we could reuse the logic.

Reply:
If we could reuse the logic, the tests in https://github.com/featurebyte/featurebyte/blob/b8ac5141588339ca221a0b952437cae810e1e176/tests/unit/service/test_feature.py will cover some of the logic here.

Reply:
hm.. let me try to refactor some of it. am also planning to refactor a bit more all at once when i've added all the logic in (for the update code paths as well), mostly wanted to try to get it to a working state first heh.